### PR TITLE
cmd/fsck: support the nlink of the check and repair directory

### LINKED ISSUE #2772: Improve `fsck`
**What would you like to be added**:
1. check nlink of a node and fix it if it's not correct
2. check recursively

**Why is this needed**:

### PR DESCRIPTION
**Pull request by zhijian-pro**
close #2772

### CODE REVIEW

File: cmd/fsck.go [hunk 1]
```diff
@@ -56,6 +56,10 @@ $ juicefs fsck redis://localhost --path /d1/d2 --repair`,
 				Name:  "repair",
 				Usage: "repair specified path if it's broken",
 			},
+			&cli.BoolFlag{
+				Name:  "recur",
```
Comment (SandyXSD) [inline cmd/fsck.go:60]: use ` --recursive, -r`

File: pkg/meta/sql.go [hunk 2]
```diff
@@ -2636,6 +2637,10 @@ func (m *dbMeta) doRepair(ctx Context, inode Ino, attr *Attr) syscall.Errno {
 				n.Nlink++
 			}
 		}
+		if ok, err := s.ForUpdate().Get(&node{Inode: inode}); err == nil && ok {
```
Comment (SandyXSD) [inline pkg/meta/sql.go:2640]:
```suggestion
ok, err := s.ForUpdate().Get(&node{Inode: inode})
if err == nil {
    if ok {
	_, err = s.Update(n, &node{Inode: inode})
    } else {
	err = mustInsert(s, n)
    }
}
return err
```

File: pkg/meta/base_test.go [hunk 3]
```diff
@@ -78,6 +80,8 @@ func testMeta(t *testing.T, m Meta) {
 	testCaseIncensi(t, m)
 	base.conf.ReadOnly = true
 	testReadOnly(t, m)
+	base.conf.ReadOnly = false
+	testCheckAndRepair(t, m)
```
Comment (SandyXSD) [inline pkg/meta/base_test.go:84]: put it before setting `ReadOnly = true`?

File: pkg/meta/base.go [hunk 4]
```diff
@@ -1044,32 +1045,114 @@ func (m *baseMeta) GetPaths(ctx Context, inode Ino) []string {
...
-		} else if !repair {
-			logger.Warnf("Path %s (inode %d) can be repaired, please re-check with 'repair' enabled", p, inode)
-		} else { // repair directory inode
+			logger.Warnf("Path %s (inode %d type %d) attribute is missing and cannot be auto-repaired, you have to repair it manually or remove it", p, inode, attr.Typ)
+		} else {
+			logger.Warnf("Path %s (inode %d) attribute is missing but can be repaired, please re-run with '--path %s --repair' to fix it", p, inode, p)
+		}
+	}
+
+	if st = m.walk(ctx, inode, fpath, &attr, func(ctx Context, inode1 Ino, path string, attr *Attr, st1 syscall.Errno) syscall.Errno {
```
Comment (SandyXSD) [inline pkg/meta/base.go:1131]: Can we do the check in parallel in recursive mode?

**Review by SandyXSD — CHANGES_REQUESTED** [review]

File: pkg/meta/redis.go [hunk 5]
```diff
@@ -2837,7 +2837,8 @@ func (m *redisMeta) doRepair(ctx Context, inode Ino, attr *Attr) syscall.Errno {
 				attr.Nlink++
 			}
 		}
-		return tx.Set(ctx, m.inodeKey(inode), m.marshal(attr), 0).Err()
+		err = tx.Set(ctx, m.inodeKey(inode), m.marshal(attr), 0).Err()
```
Comment (davies) [inline pkg/meta/redis.go:2840]: ?

Comment (zhijian-pro) [reply pkg/meta/redis.go:2840]: sorry.

**Pull request merged**